src/MiddlewareListener.php

```diff
@@ -115,53 +111,80 @@ public function onDispatch(MvcEvent $event)
+            $middlewarePipeSpec = $this->middlewareFromContainer($container, $middlewarePipeSpec);
+        }
+        if ($middlewarePipeSpec instanceof RequestHandlerInterface) {
+            return $middlewarePipeSpec;
+        }
+
+        if ($middlewarePipeSpec instanceof MiddlewareInterface) {
+            $middlewarePipeSpec = new PipeSpec($middlewarePipeSpec);
+        } elseif (! $middlewarePipeSpec instanceof PipeSpec) {
+            throw new InvalidMiddlewareException(sprintf(
+                'Route match parameter "middleware" must be one of: string container id, %s, %s or %s; %s given',
+                MiddlewareInterface::class,
+                RequestHandlerInterface::class,
+                PipeSpec::class,
+                is_object($middlewarePipeSpec) ? get_class($middlewarePipeSpec) : gettype($middlewarePipeSpec)
+            ));
+        }
+
+        // Pipe has implicit empty pipeline handler
         $pipe = new MiddlewarePipe();
-        $pipe->setResponsePrototype($responsePrototype);
-        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+        foreach ($middlewarePipeSpec->getSpec() as $middlewareToBePiped) {
             if (null === $middlewareToBePiped) {
                 throw InvalidMiddlewareException::fromNull();
             }
 
-            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
+            $middlewareName = is_object($middlewareToBePiped)
+                ? get_class($middlewareToBePiped)
+                : gettype($middlewareToBePiped);
 
-            if (is_string($middlewareToBePiped) && $serviceLocator->has($middlewareToBePiped)) {
-                $middlewareToBePiped = $serviceLocator->get($middlewareToBePiped);
+            if (is_string($middlewareToBePiped)) {
+                $middlewareToBePiped = $this->middlewareFromContainer($container, $middlewareToBePiped);
             }
-            if (! $middlewareToBePiped instanceof MiddlewareInterface && ! is_callable($middlewareToBePiped)) {
-                throw InvalidMiddlewareException::fromMiddlewareName($middlewareName);
+
+            if ($middlewareToBePiped instanceof Closure) {
+                $pipe->pipe(new CallableMiddlewareDecorator($middlewareToBePiped));
+                continue;
             }
 
-            $pipe->pipe($middlewareToBePiped);
+            if ($middlewareToBePiped instanceof MiddlewareInterface) {
+                $pipe->pipe($middlewareToBePiped);
+                continue;
+            }
+            if ($middlewareToBePiped instanceof RequestHandlerInterface) {
+                $middlewareToBePiped = new RequestHandlerMiddleware($middlewareToBePiped);
+                $pipe->pipe($middlewareToBePiped);
+                break; // request handler will always stop pipe processing. Rest of the pipe can be ignored
+            }
+
+            throw InvalidMiddlewareException::fromMiddlewareName($middlewareName);
         }
```

Comment:
Extract the contents of this loop into a method. The method should return a boolean, where one of the values means execution of the foreach should `break`:

```php
foreach ($middlewarePipeSpec->getSpec() as $middlewareToBePiped) {
    if ($this->pipeMiddleware($middlewareToBePiped, $pipe, $container)) {
        break;
    }
}
```

Reply:
I extracted whole pipe creation logic into a separate factory. Please take a look again.